### PR TITLE
Always ask for OTP code if nullok has not been set

### PR DESCRIPTION
**Pull request by nielsbasjes**
Basic solution for #69 that works on my machine.
Note that I have not yet fixed the tests (So a `make test` will fail.)

With this when I try to log in to a non existing user

I see this on the screen
```
$ ssh foo@192.168.56.102
Give me the code: 123456
Password: 
Give me the code: 
```
and this in the logs
```
Aug 17 15:25:01 localhost sshd[14263]: Invalid user foo from 192.168.56.1
Aug 17 15:25:01 localhost sshd[14263]: input_userauth_request: invalid user foo [preauth]
Aug 17 15:25:01 localhost sshd(pam_google_authenticator)[14265]: debug: start of google_authenticator for "foo"
Aug 17 15:25:01 localhost sshd(pam_google_authenticator)[14265]: Failed to read "/var/lib/google-authenticator/foo" for "foo"
Aug 17 15:25:01 localhost sshd(pam_google_authenticator)[14265]: No secret configured for user foo, asking for code anyway.
Aug 17 15:25:01 localhost sshd[14263]: Postponed keyboard-interactive for invalid user foo from 192.168.56.1 port 50578 ssh2 [preauth]
Aug 17 15:25:03 localhost sshd(pam_google_authenticator)[14265]: Dummy password supplied by PAM. Did OpenSSH 'PermitRootLogin <anything but yes>' or some other config block this login?
Aug 17 15:25:03 localhost sshd(pam_google_authenticator)[14265]: Dummy password supplied by PAM. Did OpenSSH 'PermitRootLogin <anything but yes>' or some other config block this login?
Aug 17 15:25:03 localhost sshd(pam_google_authenticator)[14265]: Invalid verification code for foo
Aug 17 15:25:03 localhost sshd[14263]: Postponed keyboard-interactive/pam for invalid user foo from 192.168.56.1 port 50578 ssh2 [preauth]
Aug 17 15:25:04 localhost sshd[14265]: pam_unix(sshd:auth): check pass; user unknown
Aug 17 15:25:04 localhost sshd[14265]: pam_unix(sshd:auth): authentication failure; logname= uid=0 euid=0 tty=ssh ruser= rhost=192.168.56.1
Aug 17 15:25:06 localhost sshd[14263]: error: PAM: Authentication failure for illegal user foo from 192.168.56.1
Aug 17 15:25:06 localhost sshd[14263]: Failed keyboard-interactive/pam for invalid user foo from 192.168.56.1 port 50578 ssh2
```

### CODE REVIEW

File: src/pam_google_authenticator.c [hunk 1]
```diff
@@ -1561,6 +1561,8 @@ static int google_authenticator(pam_handle_t *pamh, int flags,
 
   // Read and process status file, then ask the user for the verification code.
   int early_updated = 0, updated = 0;
+
+  long hotp_counter = 0;
```
Comment (ThomasHabets) [inline src/pam_google_authenticator.c]: if a `buf==NULL` check is added to `get_hotp_counter()` then this hotp_counter can remain const, and not be inside the below conditional.

File: src/pam_google_authenticator.c [hunk 2]
```diff
@@ -1561,6 +1561,8 @@ static int google_authenticator(pam_handle_t *pamh, int flags,
...
   if ((username = get_user_name(pamh, &params)) &&
```
Comment (ThomasHabets) [inline src/pam_google_authenticator.c]: Rate limit exceeded should not allow an OTP attempt. Should be possible to just move the rate limit before the `secret=`, I think.

File: src/pam_google_authenticator.c [hunk 3]
```diff
@@ -1670,43 +1682,47 @@ static int google_authenticator(pam_handle_t *pamh, int flags,
...
-                                       secret, secretLen, code, &params)) {
-          case 0:
-            rc = PAM_SUCCESS;
-            break;
-          case 1:
-            goto invalid;
-          default:
-            break;
+      switch (check_scratch_codes(pamh, &params, secret_filename, &updated, buf, code)) {
+        case 1:
```
Comment (ThomasHabets) [inline src/pam_google_authenticator.c]: wrong indendation; have `case` on same column as `switch`.

File: src/pam_google_authenticator.c [hunk 4]
```diff
@@ -1670,43 +1682,47 @@ static int google_authenticator(pam_handle_t *pamh, int flags,
...
-        break;
-      case 0:
-        rc = PAM_SUCCESS;
-        break;
-      default:
-        break;
+              break;
+        case 0:
+          rc = PAM_SUCCESS;
+              break;
```
Comment (ThomasHabets) [inline src/pam_google_authenticator.c]: bad indentation

File: src/pam_google_authenticator.c [hunk 5]
```diff
@@ -1670,43 +1682,47 @@ static int google_authenticator(pam_handle_t *pamh, int flags,
...
+        default:
+          break;
       }
 
       break;
+      }
```
Comment (ThomasHabets) [inline src/pam_google_authenticator.c]: something wrong with indentation here too

File: src/pam_google_authenticator.c (hunk 1 above)
Comment (nielsbasjes) [inline src/pam_google_authenticator.c]: Ok

File: src/pam_google_authenticator.c (hunk 3 above)
Comment (nielsbasjes) [inline src/pam_google_authenticator.c]: Yes, intendation got messed up. Fixing it.

File: .idea/codeStyleSettings.xml [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ThomasHabets) [inline .idea/codeStyleSettings.xml]: This deserves a separate commit, but I can do the commit-splitting later if you prefer.

This is for IntelliJ?

**Comment by nielsbasjes** [comment]
I also changed the noskewadj test because that was not using any of the set properties (use_first_pass, try_first_pass, etc.).

**Comment by nielsbasjes** [comment]
P.S. Please squash these commits before putting this in. Thanks.

**Pull request merged**